client/containers/clusters/index.js

```diff
@@ -2,176 +2,280 @@
+    }
+
+    /**
+     * Get the active cluster ARN.
+     */
+
+    getActiveClusterArn() {
+        if (this.state.activeClusterArn) {
+            return this.state.activeClusterArn;
+        }
+
+        const clusterName = this.props.params.clusterName;
+        const cluster = this.findCluster(clusterName);
+        // TODO: if no matching cluster, show an error
+        return cluster && cluster.clusterArn;
+    }
+
+    /**
+     * Set the current search term.
+     */
+
+    setSearchTerm(str) {
+        this.setState({ searchTerm: str });
+    }
+
+    /**
+     * Set the active cluster ARN.
+     */
+
+    setActiveCluster(cluster) {
+        this.setState({
+            activeClusterArn: cluster && cluster.clusterArn
+        });
+    }
+
+    /**
+     * Render the service sheet.
+     */
+
+    renderServiceSheet() {
+        const clusterName = this.props.params.clusterName;
+        const serviceName = this.props.params.serviceName;
+        if (!serviceName) return null;
+        const service = this.findService(clusterName, serviceName);
+        if (!service) return null;
+        // TODO: if no matching service is found, show an error
+        const cluster = this.findCluster(clusterName);
+        return <Service service={service} cluster={cluster} />
+    }
+
+    /**
+     * Render the service sheet.
+     */
+
+    renderContainerInstanceSheet() {
+        const clusterName = this.props.params.clusterName;
+        const instanceArn = this.props.params.instanceArn;
+        if (!instanceArn) return null;
+        const containerInstance = this.findContainerInstance(clusterName, instanceArn);
+        if (!containerInstance) return null;
+        // TODO: if no matching container instance is found, show an error
+        const cluster = this.findCluster(clusterName);
+        return <ContainerInstance containerInstance={containerInstance} cluster={cluster} />
+    }
+
+    /**
+     * Render the error.
+     */
+
+    renderError() {
+        return this.state.error
+            ? <ErrorMessage error={this.state.error} />
+            : null;
+    }
+
+    /**
+     * Get cluster by its `name`.
+     */
+
+    findCluster(name) {
+        return this.state.clusters.find((cluster) => {
+            return cluster.clusterName == name;
+        });
+    }
+
+    /**
+     * Get service by its `clusterName` and `serviceName`.
+     */
+
+    findService(clusterName, serviceName) {
+        const cluster = this.findCluster(clusterName);
+        if (!cluster) {
+            return;
+        }
+
+        return this.state.services.find((service) => {
+            return service.clusterArn === cluster.clusterArn &&
+                service.serviceName === serviceName;
+        });
+    }
+
+    /**
+     * Get container instance by its `clusterName` and `containerInstanceArn`.
+     */
+
+    findContainerInstance(clusterName, containerInstanceArn) {
+        const cluster = this.findCluster(clusterName);
+        if (!cluster) {
```

Comment:
Just curious what alternative, if any there is to returning null here? What would happen if these methods threw errors instead of checking for nulls?